### PR TITLE
Implement tuple<->array convertions via `From`

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This PR adds the following impls that convert between homogeneous tuples and arrays of the corresponding lengths:
```rust
impl<T> From<[T; 1]> for (T,) { ... }
impl<T> From<[T; 2]> for (T, T) { ... }
/* ... */
impl<T> From<[T; 12]> for (T, T, T, T, T, T, T, T, T, T, T, T) { ... }

impl<T> From<(T,)> for [T; 1] { ... }
impl<T> From<(T, T)> for [T; 2] { ... }
/* ... */
impl<T> From<(T, T, T, T, T, T, T, T, T, T, T, T)> for [T; 12] { ... }
```

IMO these are quite uncontroversial but note that they are, just like any other trait impls, insta-stable.

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

File: library/core/src/tuple.rs [hunk 1]
```diff
@@ -125,4 +145,24 @@ macro_rules! last_type {
     ($a:ident, $($rest_a:ident,)+) => { last_type!($($rest_a,)+) };
 }
 
+macro_rules! count {
+    ($($a:ident)*) => {
+        0 $(${ignore(a)} + 1)*
+    };
+}
+
```
Comment (klensy) [inline library/core/src/tuple.rs:153]: macro metavar ${count(x)}

**Comment by WaffleLapkin** [comment]
I've removed `()` <-> `[]` impls because they break inference in the following tests:
- `ui/hygiene/hir-res-hygiene.rs`
- `ui/suggestions/type-ascription-instead-of-path-2.rs`
- `ui/issues/issue-32709.rs`

They all basically do the following:
```rust
fn f() -> Result<(), ()> {
    Ok(())?; // <--
    Ok(())
}
```
For this to work you need to infer `?T` in `Result<(), ?T>` having only `(): From<?T>`. Currently this works because the only `impl From for ()` is the identity impl, so rustc picks it up and infers `?T = ()`.

IIRC this is not considered a breaking change per rust rules, because it can be fixed by adding types:
```rust
fn f() -> Result<(), ()> {
    Ok::<_, ()>(())?; 
    //^^^^^^^^^
    Ok(())
}
```

However I'm not sure how widespread this pattern is and if this is worth the potential breakage.

**Comment by camelid** [comment]
ping from triage: What's the status of this PR?

**Comment by scottmcm** [comment]
Looks like this is old enough that it pre-dates ACPs, so I'll nominate it for libs-api to say what they want.

**Comment by WaffleLapkin** [comment]
@camelid pretty much what the tags say: this is waiting on a `T-libs-api` decision

**Comment by m-ou-se** [comment]
We briefly discussed this in the libs-api meeting, and we were wondering if there are any motivating use cases.

**Comment by CryZe** [comment]
From the top of my head I can't really recall too many specific examples, but I definitely needed this a couple of times. In particular when dealing with different crates where one crate returns colors as (r, g, b, a) but another wants [r, g, b, a]. Similarly for xyz coordinates.

**Comment by WaffleLapkin** [comment]
So, I couldn't find the exact use case which caused this PR, going through the chat logs of the time I saw this:
```
the trait `std::convert::From<(usize, usize)>` is not implemented for `[_; 2]`
```
But I'm not sure what caused this / what was the context.

I would think that the use-case was similar to what @CryZe described — using multiple APIs some of which prefer tuples and some of which prefer arrays.

-----

@scottmcm I think that using `From` here _is_ valuable here, since it can be quite often seen on the API boundary. `.into_tuple()` and `.into_array()` would be fine too, although I would still prefer `From` (the `()` <-> `[]` as a method seems pointless to me, you are just better off writing it by hand, I think).

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by joshtriplett** [comment]
When we discussed this in today's @rust-lang/libs-api meeting, we were happy to see it merged.

But also, we wanted to track down whatever issue is causing the spurious suggestions about `From<T>` as visible in the tests this updates. Note that this issue is *not* new to these new impls, and already existed with other spuriously suggested impls. Could someone please file an issue for that, or point to an existing one if it's already filed?

Also, separate from that, suggestions should be versionsorted (1, 2, 3), not alphabetically sorted (10, 11, 12, 1, 2, ...).

**Comment by ChrisDenton** [comment]
This has passed FCP and the implementation looks good to me. The only thing I would ask for is some tests that exercise this properly. Especially because this is insta stable. Maybe doc tests? Some documentation wouldn't hurt either. E,g,, [here](https://doc.rust-lang.org/nightly/std/primitive.tuple.html#trait-implementations-1),

**Comment by ChrisDenton** [comment]
Thanks!

@bors r+

File: library/core/src/tuple.rs [hunk 2]
```diff
@@ -100,6 +100,26 @@ macro_rules! tuple_impls {
                 }
             }
         }
+
+        #[stable(feature = "array_tuple_conv", since = "1.63.0")]
```
Comment (klensy) [inline library/core/src/tuple.rs:104]: Since looks wrong here and lower?

Comment (ChrisDenton) [reply library/core/src/tuple.rs:104]: Oh, indeed! This should be using the new `CURRENT_RUSTC_VERSION` placeholder.

Comment (WaffleLapkin) [reply library/core/src/tuple.rs:104]: The PR was made before `CURRENT_RUSTC_VERSION`, oops 😅 

I'll make a fixup PR

Comment (WaffleLapkin) [reply library/core/src/tuple.rs:104]: Opened #111213

**Pull request merged**